commit c5f3529ccdb861dfd7ece92ca32d6f02f46ee4b5
Author: Michael Still <mikal@stillhq.com>
Date:   Mon Jul 7 05:37:46 2025 +1000

    Implement USB controller extra spec for libvirt.
    
    This is the implementation for the USB controller extra spec as
    desired by the new libvirt spice-direct console mode. USB device
    redirection support is a frequently requested feature for VDI users.
    
    Change-Id: I71edd03b5c63a8028c23a746c01c59d303994144

diff --git a/nova/api/validation/extra_specs/hw.py b/nova/api/validation/extra_specs/hw.py
index c50355b..40a815f 100644
--- a/nova/api/validation/extra_specs/hw.py
+++ b/nova/api/validation/extra_specs/hw.py
@@ -552,6 +552,32 @@ feature_flag_validators = [
             'enum': fields.SoundModelType.ALL,
         },
     ),
+    base.ExtraSpecValidator(
+        name='hw:usb_model',
+        description=(
+            'The model of the attached USB controller device. '
+            'Only supported by the libvirt virt driver. '
+            'If unset, no USB controller device is attached.'
+        ),
+        value={
+            'type': str,
+            'description': 'A USB controller model',
+            'enum': fields.USBControllerModelType.ALL,
+        },
+    ),
+    base.ExtraSpecValidator(
+        name='hw:redirected_usb_ports',
+        description=(
+            'The number of redirected USB ports to add to the virtual machine. '
+            'Only supported by the libvirt virt driver. If unset, no '
+            'redirected USB ports are added. The maximum value is 15.'
+        ),
+        value={
+            'type': int,
+            'description': 'The number of USB redirection devices to add',
+            'min': 0,
+        },
+    ),
 ]
 
 ephemeral_encryption_validators = [
diff --git a/nova/compute/api.py b/nova/compute/api.py
index d13a71f..a291610 100644
--- a/nova/compute/api.py
+++ b/nova/compute/api.py
@@ -125,6 +125,7 @@ MIN_COMPUTE_VDPA_HOTPLUG_LIVE_MIGRATION = 63
 SUPPORT_SHARES = 67
 
 MIN_COMPUTE_SOUND_MODEL_TRAITS = 69
+MIN_COMPUTE_USB_MODEL_TRAITS = 70
 
 # FIXME(danms): Keep a global cache of the cells we find the
 # first time we look. This needs to be refreshed on a timer or
diff --git a/nova/objects/service.py b/nova/objects/service.py
index 60863e5..5903ed7 100644
--- a/nova/objects/service.py
+++ b/nova/objects/service.py
@@ -31,7 +31,7 @@ LOG = logging.getLogger(__name__)
 
 
 # NOTE(danms): This is the global service version counter
-SERVICE_VERSION = 69
+SERVICE_VERSION = 70
 
 
 # NOTE(danms): This is our SERVICE_VERSION history. The idea is that any
@@ -240,6 +240,9 @@ SERVICE_VERSION_HISTORY = (
     # Version 69: Compute RPC v6.4:
     # Compute manager supports sound model traits
     {'compute_rpc': '6.4'},
+    # Version 70: Compute RPC v6.4:
+    # Compute manager supports USB controller model traits
+    {'compute_rpc': '6.4'},
 )
 
 # This is the version after which we can rely on having a persistent
diff --git a/nova/tests/unit/objects/test_image_meta.py b/nova/tests/unit/objects/test_image_meta.py
index 508ed4d..fab3ae8 100644
--- a/nova/tests/unit/objects/test_image_meta.py
+++ b/nova/tests/unit/objects/test_image_meta.py
@@ -569,6 +569,21 @@ class TestImageMetaProps(test.NoDBTestCase):
         primitive = obj.obj_to_primitive('1.27')
         self.assertNotIn('hw_sound_model', primitive['nova_object.data'])
 
+    def test_obj_make_compatible_usb_model_and_redirected_ports(self):
+        """Check if we pop hw_usb_model and hw_redirected_usb_ports."""
+        obj = objects.ImageMetaProps(
+            hw_usb_model='qemu-xhci',
+            hw_redirected_usb_ports=3
+        )
+        primitive = obj.obj_to_primitive()
+        self.assertIn('hw_usb_model', primitive['nova_object.data'])
+        self.assertIn('hw_redirected_usb_ports', primitive['nova_object.data'])
+
+        primitive = obj.obj_to_primitive('1.27')
+        self.assertNotIn('hw_usb_model', primitive['nova_object.data'])
+        self.assertNotIn(
+            'hw_redirected_usb_ports', primitive['nova_object.data'])
+
     def test_obj_make_compatible_socket_policy(self):
         obj = objects.ImageMetaProps(
             hw_pci_numa_affinity_policy=fields.PCINUMAAffinityPolicy.SOCKET)
diff --git a/nova/virt/hardware.py b/nova/virt/hardware.py
index d35d17f..a365440 100644
--- a/nova/virt/hardware.py
+++ b/nova/virt/hardware.py
@@ -2940,3 +2940,57 @@ def get_sound_model(
         )
 
     return model
+
+
+def get_usb_model(
+    flavor: 'objects.Flavor',
+    image_meta: 'objects.ImageMeta',
+) -> ty.Optional[str]:
+    """Get the USB controller model, if any.
+
+    :param flavor: ``nova.objects.Flavor`` instance
+    :param image_meta: ``nova.objects.ImageMeta`` instance
+    :raises: nova.exception.FlavorImageConflict if a value is specified in both
+        the flavor and the image, but the values do not match
+    :raises: nova.exception.Invalid if a value or combination of values is
+        invalid
+    :returns: A string containing the USB controller model, else None.
+    """
+    model = _get_unique_flavor_image_meta('usb_model', flavor, image_meta)
+    if model and model not in fields.USBControllerModelType.ALL:
+        raise exception.Invalid(
+            'Invalid USB controller model %(model)r. '
+            'Allowed values: %(valid)s.'
+            % {
+                'model': model,
+                'valid': ', '.join(fields.USBControllerModelType.ALL)
+            })
+
+    return model
+
+
+def get_redirected_usb_ports(
+    flavor: 'objects.Flavor',
+    image_meta: 'objects.ImageMeta',
+) -> int:
+    """Get the number of redirected USB ports, if any.
+
+    :param flavor: ``nova.objects.Flavor`` instance
+    :param image_meta: ``nova.objects.ImageMeta`` instance
+    :raises: nova.exception.FlavorImageConflict if a value is specified in both
+        the flavor and the image, but the values do not match
+    :raises: nova.exception.Invalid if a value or combination of values is
+        invalid
+    :returns: An integer number of ports, else 0.
+    """
+    count = _get_unique_flavor_image_meta(
+        'redirected_usb_ports', flavor, image_meta)
+    if not count:
+        count = 0
+    elif count > 15:
+        # NOTE(mikal): XHCI controllers only support up to 15 ports. This isn't
+        # documented in the libvirt domain XML documentation at the moment, but
+        # is at https://www.kraxel.org/blog/2018/08/qemu-usb-tips/.
+        raise exception.Invalid('Nova  only supports up to 15 USB ports.')
+
+    return count
diff --git a/nova/virt/libvirt/config.py b/nova/virt/libvirt/config.py
index fbb63a6..7b31bd2 100644
--- a/nova/virt/libvirt/config.py
+++ b/nova/virt/libvirt/config.py
@@ -4048,3 +4048,19 @@ class LibvirtConfigGuestSound(LibvirtConfigObject):
         meta = self._new_node('sound')
         meta.set('model', str(self.model))
         return meta
+
+
+class LibvirtConfigGuestUSBRedirect(LibvirtConfigObject):
+
+    def __init__(self, **kwargs):
+        super(LibvirtConfigGuestUSBRedirect,
+              self).__init__(root_name='redirdev')
+
+        self.type = 'spicevmc'
+        self.bus = 'usb'
+
+    def format_dom(self):
+        meta = self._new_node('redirdev')
+        meta.set('type', str(self.type))
+        meta.set('bus', str(self.bus))
+        return meta
diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index d6d5534..c585df4 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -6902,6 +6902,18 @@ class LibvirtDriver(driver.ComputeDriver):
         sound_device = vconfig.LibvirtConfigGuestSound(sound_model)
         guest.add_device(sound_device)
 
+    def _add_redirected_usb_ports(
+        self,
+        guest: vconfig.LibvirtConfigGuest,
+        flavor: 'objects.Flavor',
+        instance: 'objects.Instance',
+        image_meta: 'objects.ImageMeta',
+    ) -> None:
+        """Add redirected USB ports, if requested."""
+        count = hardware.get_redirected_usb_ports(flavor, image_meta)
+        for i in range(count):
+            guest.add_device(vconfig.LibvirtConfigGuestUSBRedirect())
+
     def _set_qemu_guest_agent(self, guest, flavor, instance, image_meta):
         # Enable qga only if the 'hw_qemu_guest_agent' is equal to yes
         if image_meta.properties.get('hw_qemu_guest_agent', False):
@@ -7415,7 +7427,7 @@ class LibvirtDriver(driver.ComputeDriver):
 
         return False
 
-    def _guest_add_usb_root_controller(self, guest, image_meta):
+    def _guest_add_usb_root_controller(self, guest, flavor, image_meta):
         """Add USB root controller, if necessary.
 
         Note that these are added by default on x86-64. We add the controller
@@ -7424,9 +7436,11 @@ class LibvirtDriver(driver.ComputeDriver):
         """
         usbhost = vconfig.LibvirtConfigGuestUSBHostController()
         usbhost.index = 0
+
         # an unset model means autodetect, while 'none' means don't add a
         # controller (x86 gets one by default)
         usbhost.model = None
+        specified_model = hardware.get_usb_model(flavor, image_meta)
         if not self._guest_needs_usb(guest, image_meta):
             archs = (
                 fields.Architecture.PPC,
@@ -7439,6 +7453,8 @@ class LibvirtDriver(driver.ComputeDriver):
                 # xml, where 'none' adds it but then disables it causing
                 # libvirt errors and the instances not being able to build
                 usbhost.model = None
+            elif specified_model:
+                usbhost.model = specified_model
             else:
                 usbhost.model = 'none'
         guest.add_device(usbhost)
@@ -7603,7 +7619,8 @@ class LibvirtDriver(driver.ComputeDriver):
         if self._guest_needs_pcie(guest):
             self._guest_add_pcie_root_ports(guest)
 
-        self._guest_add_usb_root_controller(guest, image_meta)
+        self._guest_add_usb_root_controller(guest, flavor, image_meta)
+        self._add_redirected_usb_ports(guest, flavor, instance, image_meta)
 
         self._guest_add_pci_devices(guest, instance)
 
diff --git a/releasenotes/notes/usb-controller-extra-specs-a2209e3563d18a26.yaml b/releasenotes/notes/usb-controller-extra-specs-a2209e3563d18a26.yaml
new file mode 100644
index 0000000..08ae50e
--- /dev/null
+++ b/releasenotes/notes/usb-controller-extra-specs-a2209e3563d18a26.yaml
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    The `hw:usb_model` flavor extra spec and the matching `hw_usb_model`
+    image property were added to allow the configuration of a USB controller
+    within an instance. This is useful with the new spice-direct console
+    type which supports passing through USB devices from the client to the
+    instance, such as a smart card reader. There is also an additional
+    `hw:redirected_usb_ports` / `hw_redirected_usb_ports` pair which
+    controls how many ports the USB controller has. This number will
+    vary based on the USB controller selected. The default remains no
+    USB controller, but when using the libvirt hypervisor driver you can
+    now also select from `qemu_xhci` and `nec_xhci`.
